### DISCUSSION TITLE
Mocking CerbosBlockingAdminClient for testing

### DISCUSSION THREAD

**Discussion by MarcoJanecki** [cerbos/cerbos#1870] [General]
Hello!

I am currently struggling a little bit with my unit tests. I am using the Java SDK, i.e. my problem corresponds to the CerbosBlockingAdminClient.
My goal is to write a unit test for a Java method which contains a call to an instance of the CerbosBlockingAdminClient. Thus, I want to mock the CerbosBlockingAdminClient (like I did for other methods for the checkResources API), such that I only run/test my Java Code. 

But I am facing the problem, that I could not find a wat to instantiate the returned value of the Client's function `cerbosBlockingAdminClient.addOrUpdatePolicy()`. The reason for this is the Builder class (`AddOrUpdatePolicyRequestBuilder`) being private in the package. Is there a way to construct an instance of the Builder to be able to mock the function call (and consecutive chained calls on the Builder), that I maybe missed out?

Thanks in advance!
Marco

**Comment by charithe** [comment]
Typically the way to mock external libraries is to create your own interface for the calls that you want to make from your code. Then you can swap that interface with a mock during tests. 

For example, you could create an interface with a method that takes the arguments you want to pass to Cerbos' `addOrUpdatePolicy` and use that interface throughout your code instead of using the Cerbos client directly. That gives you the ability to swap out the concrete implementation of the interface for testing purposes. A further advantage is that if the Cerbos API changes in the future, you'd only need to change your code once in the interface implementation.

**Comment by MarcoJanecki** [comment]
I think I found out why I got a problem. :)
While mocking the `AddOrUpdatePolicyRequestBuilder`, there was a definition of the Exception `ValidationException` (`io.envoyproxy.pgv`) missing.

Adding the dependency `build.buf.protoc-gen-validate:pgv-java-stub:1.0.2` assumably resolved that issue. :)
